### ISSUE TITLE
Can't run after Mono.Cecil

### ISSUE THREAD

**Issue by CreateAndInject** [jbevain/cecil#294]
[test.zip](https://github.com/jbevain/cecil/files/492801/test.zip)

AssemblyDefinition.ReadAssembly(source).Write(target);

**Comment by jbevain** [comment]
Yeah you're going to give more details, I'm not going to download and run a random test exe :)

**Comment by CreateAndInject** [comment]
Why you don't download?
If you worry about viruses, you can : Assembly.LoadFrom(target) or just add it as reference in Visual Studio,the "target" is not format well, so it will fail.

**Comment by CreateAndInject** [comment]
?

**Comment by atsushieno** [comment]
AssemblyDefinition.ReadAssembly() and Write() just work fine. If you want to have this bug triaged you should explain how you built your assembly so that cecil devs can write reproducible test case instead of checking in random possibly harmful binary that cecil devs cannot force cecil users to trust.

**Comment by CreateAndInject** [comment]
The source file is normal, and the target is normal by dnlib, but the target is not format well by Mono.Cecil.

**Comment by CreateAndInject** [comment]
This file is dumped from memory, but this is not impormat, because it's runnable and have been unpacked, and the target by Mono.Cecil is not format well, and it will become normal by ildasm/ilasm.it means all method codes are normal, just the file header has error.

**Comment by atsushieno** [comment]
It is important as I already explained above.

**Comment by CreateAndInject** [comment]
@atsushieno  Why are you so obstinate?

@jbevain   I have found this bug:

If PublicKey is null but Cor20HeaderFlags has StrongNameSigned, the StrongNameSignature RVA and Size can not be 0.
So, if I modify Cor20HeaderFlags from 0xb to 0x3 or keep the flags but modify StrongNameSignature RVA from 0 to 0x2050, and modify StrongNameSignature Size from 0 to 0x80, the target can run normally.

**Comment by CreateAndInject** [comment]
Mono.Cecil.PE/ImageWriter.cs line 777:
if (public_key.IsNullOrEmpty ())
				return 0;

====>

if (public_key.IsNullOrEmpty())
return (module.Attributes & ModuleAttributes.StrongNameSigned) != 0 ? 128 : 0;

**Comment by atsushieno** [comment]
Are you logically describing me as "obstinate" or is that just insulting?

**Comment by CreateAndInject** [comment]
The source is format well and the output is not format well, so if you test it, you will find the bug easylly.
But you say "AssemblyDefinition.ReadAssembly() and Write() just work fine", so I belive you don't do anything.

**Comment by CreateAndInject** [comment]
It's not important how Ibuilt the assembly.
Create a WinForm project by Visual Studio.
Modify by CFF Explorer:
.Net Directory/Flags -> 0xB
.Net Direcotry/StringNameSignature RVA -> 0x2050
.Net Direcotry/StringNameSignature Size-> 0x80

The file can run also.

after "ReadAssembly/Write", the output it not format well.

**Comment by jbevain** [comment]
@CreateAndInject thanks for looking into it. Could you please submit a PR with the fix and a test?

What @atsushieno means is that everyone is busy, if you want developers to look into your issue, you need to help them. That means putting together a better description, describing steps about how we could reproduce the issue, and that includes how to create a binary such as the one you uploaded. I'm all for brevity, but politeness also goes a long way. Just adding `?` like you're entitled to a fix or an answer is just rude.
